### PR TITLE
Continued fine-tuning template code

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
In order to showcase fine-tuning from a checkpoint, we update the fine-tuning template with some code.

For reference, the following two runs where the final ones that illustrations in this PR where made from:

1)
https://wandb.ai/anyscale-llm-forge/llmforge/runs/9c5d0_00000

s3://anyscale-production-data-cld-ldm5ez4edlp7yh4yiakp2u294w/org_4snvy99zwbmh4gbtk64jfqggmj/cld_ldm5ez4edlp7yh4yiakp2u294w/artifact_storage/artur__niederfahrenhorst/llmforge-finetuning/meta-llama/Meta-Llama-3-8B-Instruct/TorchTrainer_2024-05-28_08-04-41/TorchTrainer_9c5d0_00000_0_2024-05-28_08-04-41/checkpoint_000002

2)
https://wandb.ai/anyscale-llm-forge/llmforge/runs/3c6d2_00000

S3://anyscale-production-data-cld-ldm5ez4edlp7yh4yiakp2u294w/org_4snvy99zwbmh4gbtk64jfqggmj/cld_ldm5ez4edlp7yh4yiakp2u294w/artifact_storage/artur__niederfahrenhorst/llmforge-finetuning/meta-llama/Meta-Llama-3-8B-Instruct/TorchTrainer_2024-05-28_08-16-19/TorchTrainer_3c6d2_00000_0_2024-05-28_08-16-19/checkpoint_000001

### CODE REVIEW

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+## How to fine-tune from a previous checkpointing
+
+
+```python
+# [Optional] You can set the WandB API key to track model performance
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Continue LoRA fine-tuning on the GSM8k dataset with Llama 3 8B
+!llmforge dev finetune llama-3-8b.yaml
+```
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:21]: This is a cook book now. It has to talk about the parameters that let user control this workflow. In this section you need to talk about the two parameters and show a snippet of the yaml. 

Then say what llama-3-8b.yaml will do, if they run that command.

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+## What we are fine-tuning on
+
+Running the above command will fine-tune on the [GSM8k dataset](https://huggingface.co/datasets/gsm8k).
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:25]: In here show some wandb charts that shows that the continuation of finetuning from the second half is actually better than starting from scratch.

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+Such loss values depend greatly on the task at hand - a difference of 0.0218 may be a big improvement on some tasks and a minor improvement on others.
+
+We advise to monitor training loss and evaluation loss of fine-tunes to find out if you are improving through the continued fine-tuning.
+
+## FAQ
+
+### In what order should I fine-tune?
+
+In general: Finish with the dataset that is closest to what you want during inference.
+If you are extending the context of the model beyond it's native context length, you should start with the smallest context length end with the largest.
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:49]:
```suggestion
In general, make sure the sequence of the fine-tuning is such that you end with the dataset that is closest to what you want during inference. If you are extending the context of the model beyond its native context length, you should start with the smallest context length end with the largest.
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+### Should I extend the dataset samples or replace them with new ones when I continue fine-tuning
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:51]:
```suggestion
### Should I extend the dataset samples or replace them with new ones when I continue fine-tuning?
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+This depends on your task and how many epochs have already been trained. If in doubt, you can always watch the training and evaluation loss to see if you are overfitting.
+
+### How can I fine-tune a model that I fine-tuned on Anyscale Endpoints?
+
+You have to download the model weights through Anyscale Endpoints, upload them to a bucket of your choice and reference the bucket as an initial checkpoint in the training config yaml.
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:57]: Put this Q on the top

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+Don't forget to change the `ADAPTER_NAME` variable to fit the name of your model.
+
+# How to use this for your own purpose
+
+The training and evaluation loss of the second, the continued, fine-tuning are lower than what we saw in the first run.
+For example, the checkpoint that you find in the llama-3-8b.yaml has an evaluation loss of 0.8886.
+After continued fine-tuning, we reach a checkpoint with an evaluation loss of 0.8668.
+Such loss values depend greatly on the task at hand - a difference of 0.0218 may be a big improvement on some tasks and a minor improvement on others.
+
+We advise to monitor training loss and evaluation loss of fine-tunes to find out if you are improving through the continued fine-tuning.
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:42]: I would write this part differently. 

1. I would talk about the diffs in terms of loss value via showing the diffs in the loss curve and zooming / annotating the parts that makes it super clear how continued ft helped.
2. I would not get into the details of loss values themselves, just show the curves.
3. Are there differences in final accuracy when you do the two? If not what is the value of showing how to evaluate? I would just entirely remove that as we have already proved the point via showing the loss curves. Unless there is a clear gap in the final performance, there is no need to confused reader with evaluation code.

**Comment by kouroshHakha** [comment]
Also show how a checkpoint repository should look like for LoRA and full parameter when you describe the parameters.

File: templates/batch-llm/README.md [hunk 7]
```diff
@@ -223,7 +223,7 @@ ds = ds.map_batches(
 
 Finally, make sure to either enable *Auto-select worker nodes* or configure your workspace cluster to have the appropriate GPU worker nodes (A10G or L4):
 
-<img src="https://raw.githubusercontent.com/anyscale/templates/main/templates/batch-llm/assets/ray-data-gpu.png"/>
+![title](assets/ray-data-gpu.png)
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply templates/batch-llm/README.md]: No and I can't remember making that change. Likely was done by the linter script which creates READMEs from notebooks in this case.

**Review by kouroshHakha — COMMENTED** [review]
I think all the changes you made are not actually commited because you changed them on the README.md file and when you commit README.ipynb gets converted to README.md therefore overriding all your changes.

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md]: 1. Ok I fixed this. The x- and y-axis of both runs are now uniform across the to pics to give a better perspective on how the losses related. I've also annotated the curves with their values to make it crystal clear what's happening.
2. I got rid of that whole part.
3. I got rid of the eval script.

Thanks for the feedback!

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+**⏱️ Time to complete**: 40 minutes
+
+This guide assumes that you have familiarized yourself with the main fine-tuning guide of this template.
+In this cookbook tutorial, we showcase how a checkpoint that was created earlier can be used as initialization for another round of fine-tuning.
+This allows us to sequentially combine fine-tuning on multiple datasets in order to get performance boost on the final task that we care about. 
+
+We support both Full-parameter checkpoints, and LoRA-adapter checkpoints. However, we recommend not combining the two by training a full-parameter model followed by a LoRA adaptation, because serving the resulting LoRA adapter will require the full-parameter checkpoint. Unless you are fine-tuning many such LoRA adaptors for different tasks this serving architecture does not have the neither the economical benefits of LoRA nor the quality benefits of full-parameter.
+
+## How to fine-tune from a previous checkpointing
+
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:12]:
```suggestion
To get started, we can run the following illustrative example:
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Continue LoRA fine-tuning on the GSM8k dataset with Llama 3 8B
+!llmforge dev finetune llama-3-8b.yaml
+```
+
+## What we are fine-tuning on
+
+Running the above command will fine-tune on the [GSM8k dataset](https://huggingface.co/datasets/gsm8k). 
+In this example, we split the dataset into two halfs, each consisting of approximately 4.000 samples.
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:25]:
```suggestion
In this example, we have splited the dataset into two halves, each consisting of approximately 4,000 samples.
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+```python
+# [Optional] You can set the WandB API key to track model performance
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Continue LoRA fine-tuning on the GSM8k dataset with Llama 3 8B
+!llmforge dev finetune llama-3-8b.yaml
+```
+
+## What we are fine-tuning on
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:22]: Let's remove this title and continue explaining what happens in the code-block in the next paragraph.

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+Running the above command will fine-tune on the [GSM8k dataset](https://huggingface.co/datasets/gsm8k). 
+In this example, we split the dataset into two halfs, each consisting of approximately 4.000 samples.
+The provided initial checkpoint has been trained on the first half and is already good at solving GSM8k. By running the above command, you continue fine-tuning from the provided checkpoint with the second half.
+
+Note the following evaluation losses. The first graph shows the evaluation loss on three epochs of training on the first half of the GSM8k dataset.
+
+<img src="https://raw.githubusercontent.com/anyscale/templates/main/templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/../../assets/3epochs_1st_dataset.png" alt="evaluation loss of 1st training" width="700"/>
+
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:31]: Is it possible to overlay the two images? and add legends? instead of two separate curves?

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    "# initial_base_model_ckpt_path: ...\n",
+    "initial_adapter_model_ckpt_path: s3://large-dl-models-mirror/finetuning_template/continued_ft_gsm8k_checkpoint\n",
+    "train_path: s3://large-dl-models-mirror/finetuning_template/train_2.jsonl\n",
+    "# ...\n",
+    "```\n",
+    "\n",
+    "We fine-tune Llama 3 8B Instruct, but the initial weights of the LoRA adapter are loaded from our s3 mirror.\n",
+    "It makes sense to keep those weights in a bucket so that they can be accessed from all nodes of your cluster.\n",
+    "The train path `(.../train_2.jsonl)` points to the second part of the GSM8k dataset that we fine-tune on.\n",
+    "If we were not fine-tuning with LoRA, we would not configure `initial_adapter_model_ckpt_path`, but `initial_base_model_ckpt_path` instead.\n",
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb:69]:
```suggestion
    "If we wanted to continue the finetuning of a full-parameter checkpoint, we should configure `initial_base_model_ckpt_path` instead of `initial_adapter_model_ckpt_path`. \n",
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    "Such loss values depend greatly on the task at hand - a difference of 0.0218 may be a big improvement on some tasks and a minor improvement on others.\n",
+    "\n",
+    "We advise to monitor training loss and evaluation loss of fine-tunes to find out if you are improving through the continued fine-tuning.\n",
+    "\n",
+    "## FAQ\n",
+    "\n",
+    "### In what order should I fine-tune?\n",
+    "\n",
+    "In general: Finish with the dataset that is closest to what you want during inference.\n",
+    "If you are extending the context of the model beyond it's native context length, you should start with the smallest context length end with the largest.\n",
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb:85]:
```suggestion
    "If you are extending the context of the model beyond its native context length, you should start with the smallest context length end with the largest.\n",
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    "\n",
+    "### Should I extend the dataset samples or replace them with new ones when I continue fine-tuning\n",
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb:87]:
```suggestion
    "### Should I extend the dataset samples or replace them with new ones when I continue fine-tuning?\n",
```

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    "\n",
+    "This depends on your task and how many epochs have already been trained. If in doubt, you can always watch the training and evaluation loss to see if you are overfitting.\n",
+    "\n",
+    "### How can I fine-tune a model that I fine-tuned on Anyscale Endpoints?\n",
+    "\n",
+    "You have to download the model weights through Anyscale Endpoints, upload them to a bucket of your choice and reference the bucket as an initial checkpoint in the training config yaml."
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.ipynb:93]: Can we add a screen shot on how to download, where to click?

**Review by kouroshHakha — COMMENTED** [review]
Some nits are left.

File: templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md (hunk 10 above)
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:22]: You mean remove the entire block of "What are we fine-tuning on"?

Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/README.md:24]: No just the title.

File: templates/fine-tune-llm_v2/assets/3epochs_1st_dataset.png
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/assets/3epochs_1st_dataset.png:1]: can we get all these files under `assets/cookbooks/conitune_...`

File: templates/fine-tune-llm_v2/README.md [hunk 16]
```diff
@@ -63,6 +63,12 @@ Best checkpoint is stored in:
 
 Follow the [Learn how to bring your own models](https://docs.anyscale.com/examples/deploy-llms#more-guides) section under the llm serving template to serve this fine-tuned model with the specified storage uri.
 
+## Finally
+
+After you are with the above, you can find recipies that extend the functionality of this template under the cookbooks folder:
+
+* (Continue fine-tuning from a previous checkpoint)[cookbooks/continue_from_checkpoint/README.md]
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/README.md:70]:
```suggestion
* [Continue fine-tuning from a previous checkpoint](cookbooks/continue_from_checkpoint/README.md)
```

File: templates/fine-tune-llm_v2/README.ipynb [hunk 17]
```diff
@@ -78,6 +78,12 @@
     "\n",
     "Follow the [Learn how to bring your own models](https://docs.anyscale.com/examples/deploy-llms#more-guides) section under the llm serving template to serve this fine-tuned model with the specified storage uri.\n",
     "\n",
+    "## Finally\n",
+    "\n",
+    "After you are with the above, you can find recipies that extend the functionality of this template under the cookbooks folder:\n",
+    "\n",
+    "* (Continue fine-tuning from a previous checkpoint)[cookbooks/continue_from_checkpoint/README.md]\n",
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/README.ipynb:85]:
```suggestion
    "* [Continue fine-tuning from a previous checkpoint](cookbooks/continue_from_checkpoint/README.md)\n",
```

**Review by kouroshHakha — COMMENTED** [review]
a few things left:

**Review by kouroshHakha — APPROVED** [review]
